File: backend-spring-boot/src/main/java/no/vicx/backend/user/validation/AtLeastOneNotNull.java
package no.vicx.backend.user.validation;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Constraint(validatedBy = AtLeastOneNotNullValidator.class)
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
public @interface AtLeastOneNotNull {
    String message() default "{vicx.constraints.AtLeastOneNotNull.message}";
    String propertyNodeName() default "patchRequestBody";
    Class<?>[] groups() default {};
    Class<? extends Payload>[] payload() default {};
}

File: backend-spring-boot/src/main/kotlin/no/vicx/backend/user/validation/AtLeastOneNotNull.kt
package no.vicx.backend.user.validation

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.reflect.KClass


@Constraint(validatedBy = [AtLeastOneNotNullValidator::class])
@Target(AnnotationTarget.CLASS)
@Retention(
    AnnotationRetention.RUNTIME
)
annotation class AtLeastOneNotNull(
    val message: String = "{vicx.constraints.AtLeastOneNotNull.message}",
    val propertyNodeName: String = "patchRequestBody",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = []
)